fpgaconvnet/models/layers/Layer.py

```diff
@@ -66,283 +55,181 @@ def __post_init__(self):
+        return self.workload_in() / self.streams_in() 
 
-    def width_out(self):
-        """
-        Returns
-        -------
-        int
-            data width out
-        """
-        return self.data_t.width
+    def size_out(self) -> int:
+        return self.workload_out() / self.streams_out() 
 
     def latency_in(self):
-        return int(abs(self.workload_in()/(min(self.mem_bw_in, self.rate_in()*self.streams_in()))))
+            """
+            Calculates the latency of input data for this layer.
+
+            Returns:
+                int: The calculated latency value.
+            """
+            return int(abs(self.workload_in()/(min(self.mem_bw_in, self.rate_in()*self.streams_in()))))
 
     def latency_out(self):
-        return int(abs(self.workload_out()/(min(self.mem_bw_out, self.rate_out()*self.streams_out()))))
+            """
+            Calculates the latency of the output of the layer.
+
+            Returns:
+                int: The latency of the output of the layer.
+            """
+            return int(abs(self.workload_out()/(min(self.mem_bw_out, self.rate_out()*self.streams_out()))))
 
     def latency(self):
-        return max(self.latency_in(), self.latency_out())
+            """
+            Returns the maximum latency of the input and output of the layer.
+
+            Returns:
+                The maximum latency of the input and output of the layer.
+            """
+            return max(self.latency_in(), self.latency_out())
 
     def pipeline_depth(self):
-        return sum([ self.modules[module].pipeline_depth() for module in self.modules ])
+            """
+            Computes the pipeline depth of the layer by summing the pipeline depths of its modules.
+
+            Returns:
+                The pipeline depth of the layer.
+            """
+            return sum([ self.modules[module].pipeline_depth() for module in self.modules ])
 
     def wait_depth(self):
         return sum([ self.modules[module].wait_depth() for module in self.modules ])
 
+    @abstractmethod
     def resource(self):
-        return {
-            "LUT"   : 0,
-            "FF"    : 0,
-            "BRAM"  : 0, #bram_array_resource_model(self.buffer_depth,self.data_t.width, "fifo")*self.streams_in(),
-            "DSP"   : 0
-        }
+        pass
 
     def memory_bandwidth(self):
         return {
             "in"  : min(self.mem_bw_in, self.rate_in()*self.streams_in()),
             "out" : min(self.mem_bw_out, self.rate_out()*self.streams_out())
         }
 
+    @abstractmethod
     def get_coarse_in_feasible(self):
-        return get_factors(int(self.channels_in()))
+        pass
 
+    @abstractmethod
     def get_coarse_out_feasible(self):
-        return get_factors(int(self.channels_out()))
+        pass
 
+    @abstractmethod
     def update(self):
         pass
 
     def layer_info(self, parameters, batch_size=1):
-        parameters.batch_size   = batch_size
```

Comment:
why are we dropping some fields?

Reply:
@AlexMontgomerie I guess these are inherited from the base LayeBase class, but is this true @AlexMontgomerie or we need to add them back?